src/OpenFeature/EventExecutor.cs

```diff
@@ -234,82 +223,92 @@ private async void ProcessFeatureProviderEventsAsync(object? providerRef)
-                                {
-                                    if (this._clientHandlers.TryGetValue(keyAndValue.Key, out var clientRegistry))
-                                    {
-                                        if (e.EventPayload?.Type != null && clientRegistry.TryGetValue(e.EventPayload.Type, out var clientEventHandlers))
-                                        {
-                                            foreach (var eventHandler in clientEventHandlers)
-                                            {
-                                                this.InvokeEventHandler(eventHandler, e);
-                                            }
-                                        }
-                                    }
-                                }
-                            }
-
-                            if (e.Provider != this._defaultProvider)
-                            {
-                                break;
-                            }
-                            // handling the default provider - invoke event handlers for clients which are not bound
-                            // to a particular feature provider
-                            foreach (var keyAndValues in this._clientHandlers)
-                            {
-                                if (this._namedProviderReferences.TryGetValue(keyAndValues.Key, out _))
-                                {
-                                    // if there is an association for the client to a specific feature provider, then continue
-                                    continue;
-                                }
-                                if (e.EventPayload?.Type != null && keyAndValues.Value.TryGetValue(e.EventPayload.Type, out var clientEventHandlers))
-                                {
-                                    foreach (var eventHandler in clientEventHandlers)
-                                    {
-                                        this.InvokeEventHandler(eventHandler, e);
-                                    }
-                                }
-                            }
-                        }
-                        break;
+                    continue;
+                }
+
+                lock (this._lockObj)
+                {
+                    this.ProcessApiHandlers(e);
+                    this.ProcessClientHandlers(e);
+                    this.ProcessDefaultProviderHandlers(e);
+                }
+            }
+        }
+
+        private void ProcessApiHandlers(Event e)
+        {
+            if (e.EventPayload?.Type != null && this._apiHandlers.TryGetValue(e.EventPayload.Type, out var eventHandlers))
+            {
+                foreach (var eventHandler in eventHandlers)
+                {
+                    this.InvokeEventHandler(eventHandler, e);
+                }
+            }
+        }
+
+        private void ProcessClientHandlers(Event e)
```

Comment:
You removed a comment, but I guess the method name makes it clear what's happening now, so I'm fine with it.